### PR TITLE
Slack 4 0 adaption

### PR DESCRIPTION
**Pull request by mosheBelostotsky**

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -1,37 +0,0 @@
-'use strict';
```
Comment (rqc) [inline Gruntfile.js]: @mosheBelostotsky just out of curiosity, why did you remove this gruntfile?

File: lib/advanced_adapter/adapter-slack.coffee [hunk 2]
```diff
@@ -59,11 +59,26 @@ class Adapter
...
+    .then (r) ->
+      opts.channel = r
+      return SlackApi.channels.info(opts)
     .then (r) ->
       return r.channel
 
+  channelNameToId: (channel) ->
+    _this = @
+    if not channel.startsWith('#')
+      return Promise.resolve(channel)
```
Comment (rqc) [inline lib/advanced_adapter/adapter-slack.coffee]: By resolving the promise at this point what is your intended behavior? Correct me if I'm wrong, but we would not like to return the `channel` if it doesn't start with `#`.

File: lib/advanced_adapter/adapter-slack.coffee [hunk 3]
```diff
@@ -59,11 +59,26 @@ class Adapter
...
+    else
```
Comment (rqc) [inline lib/advanced_adapter/adapter-slack.coffee]: we might not need this `else`

File: lib/advanced_adapter/adapter-slack.coffee [hunk 4]
```diff
@@ -59,11 +59,26 @@ class Adapter
       return r.channels
 
   channelInfo: (channelId) ->
-    opts =
-      token: @apiToken
-      channel: channelId
-    return SlackApi.channels.info(opts)
+    opts = {token: @apiToken}
```
Comment (rqc) [inline lib/advanced_adapter/adapter-slack.coffee]: Do you think it is worth adding some layer of validation at this point for the `apiToken`? Perhaps check if it is empty at constructor time or before each call...

File: lib/advanced_adapter/adapter-slack.coffee [hunk 5]
```diff
@@ -59,11 +59,26 @@ class Adapter
...
+      return SlackApi.channels.info(opts)
     .then (r) ->
       return r.channel
 
+  channelNameToId: (channel) ->
+    _this = @
+    if not channel.startsWith('#')
+      return Promise.resolve(channel)
+    else
+      return new Promise (resolve, reject) ->
```
Comment (rqc) [inline lib/advanced_adapter/adapter-slack.coffee]: Lines 75-82 are really hard to read. Maybe it is my coffeescript sensitive eyes :smile: Perhaps we could refactor at least one of this callbacks into its own function so that it looks easier to read?

File: package.json [hunk 6]
```diff
@@ -33,19 +28,21 @@
     "hubot": "2.x"
   },
   "devDependencies": {
+    "hubot-test-helper": "github:eedevops/hubot-test-helper",
```
Comment (rqc) [inline package.json]: Questions:
- Is this based on a third-party open source project?
- If so, why are we using a fork under `eedevops`?

File: package.json [hunk 7]
```diff
@@ -33,19 +28,21 @@
...
-    "matchdep": "~0.3.0",
-    "mocha": "^2.1.0",
-    "sinon": "^1.13.0",
-    "sinon-chai": "^2.7.0"
+    "jshint": "^2.9.2",
+    "coffeelint": "^1.15.7",
+    "jenkins-mocha": "^2.6.0"
+  },
+  "main": "index.coffee",
+  "scripts": {
```
Comment (rqc) [inline package.json]: These `scripts` are run sometime during npm manipulation? Could you explain a bit more how you are using them in:
- Your development environment
- The CI / CD (publishing) of the npm package

One thing to consider is, is there are way to keep this as a `gulp` or `gruntfile` recipe? What would be the pros / cons of that?

File: src/0_bootstrap.coffee [hunk 8]
```diff
@@ -77,7 +77,7 @@ module.exports = (robot) ->
     if !res
       product = if product then product.trim() else ""
       res = "\nthere is no such integration #{product}"
-    res = "enterprise help"+res
+    res = "help for hubot enterprise:"+res
```
Comment (rqc) [inline src/0_bootstrap.coffee]: Can we create a commons interface or object that can be used for these text messages. Eventually it may become hard to maintain.

File: test/enterprise-test.coffee [hunk 9]
```diff
@@ -75,7 +75,7 @@ describe 'enterprise tests', ->
     @room.user.say('alice', '@hubot enterprise').then =>
       expect(@room.messages).to.eql [
         [ 'alice', '@hubot enterprise' ],
-        [ 'hubot', '@alice enterprise help\n'+
+        [ 'hubot', '@alice help for hubot enterprise:\n'+
```
Comment (rqc) [inline test/enterprise-test.coffee]: could we create some mock objects that refactor some of these messages that are used in more than one place? That way, we avoid duplicates, as well as reducing the possibility of typos in our tests.

File: src/admin.coffee [hunk 10]
```diff
@@ -25,23 +25,17 @@ module.exports = (robot) ->
...
-      msg.envelope.message.user.room == 'general')
-        msg.reply 'cannot archive private or general channel'
-        return
-      channelId = ['', msg.envelope.message.rawMessage.channel]
+      channel = ['', msg.envelope.message.room]
     else if (!msg.match[1].startsWith('#'))
       msg.reply 'channel name should start with #'
       return
-    else if (!(channelId = /<\#(.*)>/i.exec(msg.envelope.message.rawText)))
+    else if (!(channel = /(#[^\s]+)/i.exec(msg.envelope.message.text)))
```
Comment (rqc) [inline src/admin.coffee]: Bare with me here since I am still learning hubot, slack, etc. One thing that we should make sure is that the regex is as close as possible to the documented format. I could only find this [link](https://get.slack.help/hc/en-us/articles/201654073-Renaming-a-channel) and it says that periods `.` are not allowed, which I think are included in `\s`. Also, there is a max length of `21` chars. If this is enforced by the slack api adapter, then it is fine, but we should make sure!

File: lib/advanced_adapter/adapter-slack.coffee [hunk 11]
```diff
@@ -59,11 +59,26 @@ class Adapter
...
-    return SlackApi.channels.info(opts)
+    opts = {token: @apiToken}
+    return @channelNameToId(channelId)
+    .then (r) ->
+      opts.channel = r
+      return SlackApi.channels.info(opts)
     .then (r) ->
       return r.channel
 
+  channelNameToId: (channel) ->
```
Comment (rqc) [inline lib/advanced_adapter/adapter-slack.coffee]: Did you write a new test for this adapter function?

**Comment by rqc** [comment]
@mosheBelostotsky overall it looks good. Before merging I would like you to consider the following:

- [ ] Address all the comments made in the code.
- [ ] Add descriptions for the purpose of each file in the source code. For somebody new it is hard to follow as is. 
- [ ] Add comments for adapter functions detailing its purpose. Investigate if there is a documentation generator that we could use for node modules or perhaps something specific to hubot.

Thanks!

File: package.json [hunk 12]
```diff
@@ -1,13 +1,8 @@
 {
   "name": "hubot-enterprise",
-  "version": "0.0.11",
+  "version": "0.0.12",
```
Comment (rqc) [inline package.json]: This is a tricky one. We should have semantic versioning for this integration, no doubt. However, I have a hard time performing these changes bundled in a pull request of a feature. Even more troubling is the scenario when we pull request a new feature or bugfix and forget to update the `package.json`. We need to make sure that we have a process for this that is consistent for all of us and that is error-proof. Also, I really like having software versions correlated to *git tags*. Let's say we want to debug or troubleshoot a specific version, it is easier to do so by asking or investigating the version of the package that is causing the issue, and checking out the code via a git tag from the source code. Please take a look at this https://docs.npmjs.com/cli/version in particular the part that talks about using the *git tag*, and let me know what you think.

File: Gruntfile.js (hunk 1 above)
Comment (mosheBelostotsky) [inline Gruntfile.js]: We don't need it's functionality
I moved most of it to package.json
And the rest is in CI pipeline.
Please see /eedevops/generator-hubot-enterprise

File: lib/advanced_adapter/adapter-slack.coffee (hunk 4 above)
Comment (mosheBelostotsky) [inline lib/advanced_adapter/adapter-slack.coffee]: Slack API is handling it..
But I can add on the lower layer a check.

File: lib/advanced_adapter/adapter-slack.coffee (hunk 2 above)
Comment (mosheBelostotsky) [inline lib/advanced_adapter/adapter-slack.coffee]: If the channel does not start with '#' it must be channel ID so we return it as is.
If it starts with '#' we must convert the name to id.
Most of slack API called use channel ID rather than name...

File: package.json (hunk 12 above)
Comment (mosheBelostotsky) [inline package.json]: Software version are tight to git tags.
Jenkins CI scripts creating new tag+version once recognizing new version in package.json when building MASTER branch.
We can discuss this more with @eitzhaki

File: lib/advanced_adapter/adapter-slack.coffee (hunk 11 above)
Comment (mosheBelostotsky) [inline lib/advanced_adapter/adapter-slack.coffee]: No. Not yet. I will discuss how to do it with @JonathanRosado @gmatoshp

File: package.json (hunk 6 above)
Comment (mosheBelostotsky) [inline package.json]: I manipulated the code.
I want to send pull request to them...

Comment (rqc) [inline package.json]: ok. in that case, please let's start that process for the pull req asap. We musn't rely on our fork for long.

File: package.json (hunk 12 above)
Comment (rqc) [inline package.json]: I think that's a good approach. However, to mitigate the potential issues that I have outlined before, it would be best to do implement it the other way around. Our tagging in the `package.json` should be as a result of a `git tag` and not the other way around. If you see any potential problems with this approach please let me know.

File: lib/advanced_adapter/adapter-slack.coffee (hunk 4 above)
Comment (rqc) [inline lib/advanced_adapter/adapter-slack.coffee]: ok, I'll leave it to your discretion. Thanks!

File: Gruntfile.js (hunk 1 above)
Comment (rqc) [inline Gruntfile.js]: What I would like to make sure, is that tasks like testing, linting, etc. are part of the *development* process as well, and we are not offloading everything to the CI. It is a good practice to run tasks as part of your everyday development tasks. I wonder if leaving `grunt / gulp` is the right move here.
